File: bindings/dotnet/dotnet-binding/Microsoft.Azure.IoT.Gateway/IGatewayModule.cs
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Microsoft.Azure.IoT.Gateway
{
    /// <summary> Interface to be implemented by the .NET Module </summary>
    public interface IGatewayModule
    {
        /// <summary>
        ///     Creates a module using the specified configuration connecting to the specified message broker.
        /// </summary>
        /// <param name="broker">The broker to which this module will connect.</param>
        /// <param name="configuration">A byte[] with user-defined configuration for this module. This parameter shall be enconded to a UTF-8 String.</param>
        /// <returns></returns>
        void Create(Broker broker, byte[] configuration);

        /// <summary>
        ///     Disposes of the resources allocated by/for this module.
        /// </summary>
        /// <returns></returns>
        void Destroy();

        /// <summary>
        ///     The module's callback function that is called upon message receipt.
        /// </summary>
        /// <param name="received_message">The message being sent to the module.</param>
        /// <returns></returns>                
        void Receive(Message received_message);
    }
}

File: bindings/java/gateway-java-binding/src/main/java/com/microsoft/azure/gateway/core/IGatewayModule.java
/*
 * Copyright (c) Microsoft. All rights reserved.
 * Licensed under the MIT license. See LICENSE file in the project root for full license information.
 */
package com.microsoft.azure.gateway.core;

public interface IGatewayModule {

    /**
     * The create method is called by the subclass constructor when the native Gateway creates the module. The constructor
     * should save the {@code broker} parameter.
     *
     * @param moduleAddr The address of the native module pointer
     * @param broker The {@link Broker} to which this module belongs
     * @param configuration The configuration for this module represented as a JSON string
     */
    void create(long moduleAddr, Broker broker, String configuration);

    /**
     * The start method is called when the native Gateway has created all modules to notify each module that it is safe
     * to start sending messages.
     *
     * This method is optional, but it is recommended that any long-running processes get started in this method.
     */
    void start();

    /**
     * The receive method is called on a {@link GatewayModule} whenever it receives a message.
     *
     * The destroy() and receive() methods are guaranteed to not be called simultaneously.
     *
     * @param source The serialized message
     */
    void receive(byte[] source);

    /**
     * The destroy method is called on a {@link GatewayModule} before it is about to be "destroyed" and removed from the gateway.
     * Once a module is removed from the gateway, it may no longer send or receive messages.
     *
     * The destroy() and receive() methods are guaranteed to not be called simultaneously.
     */
    void destroy();
}